### PR TITLE
Add nix support

### PR DESCRIPTION
**Pull request by jbgi**
TODO
-  [x] activate repo in bors
-  [x] hydra jobset
-  [ ] run coverage (stack-rebuild) on ci (?)
-  [ ] add docker images

### CODE REVIEW

File: nix/.stack.nix/default.nix [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        cardano-crypto-test = ./cardano-crypto-test.nix;
+        io-sim = ./io-sim.nix;
+        io-sim-classes = ./io-sim-classes.nix;
+        network-mux = ./network-mux.nix;
+        ouroboros-network = ./ouroboros-network.nix;
+        ouroboros-consensus = ./ouroboros-consensus.nix;
+        typed-protocols = ./typed-protocols.nix;
+        typed-protocols-cbor = ./typed-protocols-cbor.nix;
+        cardano-explorer-db = ./cardano-explorer-db.nix;
+        cardano-explorer-db-test = ./cardano-explorer-db-test.nix;
```
[the remark this answers is not in the document]
Comment (craigem) [reply nix/.stack.nix/default.nix]: Resolved with commit b8bdd7c

**Comment by jbgi** [comment]
bors ping

**Comment by jbgi** [comment]
bors try

File: nix/docker.nix [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  txSubmitDockerImage = let
+    clusterStatements = lib.concatStringsSep "\n" (lib.mapAttrsToList (_: value: value) (commonLib.forEnvironments (env: ''
+      elif [[ "$NETWORK" == "${env.name}" ]]; then
+        exec ${scripts.${env.name}.tx-submit}
+    '')));
+    entry-point = writeScriptBin "entry-point" ''
+      #!${runtimeShell}
+      if [[ -d /config ]]; then
+         exec ${cardano-tx-submit-webapi}/bin/cardano-tx-submit-webapi \
+           --socket-path /data/node.socket
```
Comment (cleverca22) [inline nix/docker.nix]: the `\` is missing on this line

File: nix/docker.nix [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  in dockerTools.buildImage {
+    name = "${explorerApiRepoName}";
+    fromImage = baseImage;
+    tag = "${gitrev}";
+    created = "now";   # Set creation date to build time. Breaks reproducibility
+    contents = [ entry-point ];
+    config = {
+      EntryPoint = [ "${entry-point}/bin/entry-point" ];
+      ExposedPorts = {
+        "3001/tcp" = {};  # Cardano node p2p
```
Comment (cleverca22) [inline nix/docker.nix]: explorer-api doesnt use the cardano p2p protocol, but is exposing an http api on a port, dont remember which one

File: nix/docker.nix [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  in dockerTools.buildImage {
+    name = "${txSubmitRepoName}";
+    fromImage = baseImage;
+    tag = "${gitrev}";
+    created = "now";   # Set creation date to build time. Breaks reproducibility
+    contents = [ entry-point ];
+    config = {
+      EntryPoint = [ "${entry-point}/bin/entry-point" ];
+      ExposedPorts = {
+        "3001/tcp" = {};  # Cardano node p2p
```
Comment (cleverca22) [inline nix/docker.nix]: the 3001 here is the http api of tx submitter, not the p2p protocol for nodes

File: nix/docker.nix (hunk 3 above)
Comment (rhyslbw) [reply nix/docker.nix]: 8100

File: nix/docker.nix [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+      elif [[ "$NETWORK" == "${env.name}" ]]; then
+        exec ${scripts.${env.name}.tx-submit}
+    '')));
+    entry-point = writeScriptBin "entry-point" ''
+      #!${runtimeShell}
+      if [[ -d /config ]]; then
+         exec ${cardano-tx-submit-webapi}/bin/cardano-tx-submit-webapi \
+           --socket-path /data/node.socket
+           --genesis-file /config/genesis.json \
+           --port 3001 \
```
Comment (rhyslbw) [inline nix/docker.nix]:
```suggestion
           --port 8101 \
```

File: nix/docker.nix (hunk 3 above)
Comment (rhyslbw) [inline nix/docker.nix]:
```suggestion
        "8100/tcp" = {}; 
```

File: nix/docker.nix [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      iputils           # Useful utilities for Linux networking
+      socat             # Utility for bidirectional data transfer
+      utillinux         # System utilities for Linux
+    ];
+    # set up /tmp (override with TMPDIR variable)
+    extraCommands = ''
+      mkdir -m 0777 tmp
+    '';
+  };
+  apiDockerImage = let
```
Comment (rhyslbw) [inline nix/docker.nix]:
```suggestion
  explorerApiDockerImage = let
```
Let's keep it consistent with the service name

File: nix/docker.nix (hunk 4 above)
Comment (rhyslbw) [inline nix/docker.nix]:
```suggestion
        "8101/tcp" = {};
```

File: nix/docker.nix [hunk 7]
```diff
@@ -0,0 +1,141 @@
+############################################################################
+# Docker image builder
+#
+# To build and load into the Docker engine:
+#
+#   docker load -i $(nix-build -A dockerImage.explorerApi --no-out-link)
+#   docker load -i $(nix-build -A dockerImage.txSubmit --no-out-link)
+#
+# To launch with provided mainnet configuration and persist state in a docker
```
Comment (rhyslbw) [inline nix/docker.nix]: There's no state to persist
```suggestion
# To launch with provided mainnet configuration
```

File: nix/docker.nix [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+# volume run:
+#
+#   docker run -v /data -e NETWORK=mainnet inputoutput/cardano-explorer-api:<TAG>
```
Comment (rhyslbw) [inline nix/docker.nix]:
```suggestion
#   docker run NETWORK=mainnet inputoutput/cardano-explorer-api:<TAG>
```

File: nix/docker.nix [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+#   docker run -v /data -e NETWORK=mainnet inputoutput/cardano-tx-submit:<TAG>
```
Comment (rhyslbw) [inline nix/docker.nix]:
```suggestion
#   docker run -e NETWORK=mainnet inputoutput/cardano-tx-submit:<TAG>
```

File: nix/docker.nix [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+#
+# To launch with provided testnet configuration without persisting state run:
```
Comment (rhyslbw) [inline nix/docker.nix]:
```suggestion
# To launch with provided testnet configuration
```

File: nix/docker.nix (hunk 8 above)
Comment (rhyslbw) [reply nix/docker.nix]: Cardano Explorer API isn't concerned about the network, it just connects to the DB

File: nix/docker.nix [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+#
+#   docker run -e NETWORK=testnet inputoutput/cardano-explorer-api:<TAG>
```
Comment (rhyslbw) [inline nix/docker.nix]:
```suggestion
```

File: nix/nixos/cardano-tx-submitter-service.nix [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+# notes:
+# this service exposes an http port, and connects to a cardano-node over a unix socket
+let
+  cfg = config.services.cardano-tx-submit;
+  envConfig = cfg.environment;
+  localPkgs = import ../. {};
+in {
+  options = {
+    services.cardano-tx-submit = {
+      enable = lib.mkEnableOption "enable the cardano-explorer tx submitter api";
```
Comment (rhyslbw) [inline nix/nixos/cardano-tx-submitter-service.nix]: For consistency this and the file name should be referred to as `cardano-tx-submit`. it's not an "explorer" component
```suggestion
      enable = lib.mkEnableOption "enable cardano-tx-submit";
```

File: nix/nixos/module-list.nix [hunk 13]
```diff
@@ -0,0 +1,4 @@
+[
+  ./cardano-explorer-api-service.nix
+  ./cardano-tx-submitter-service.nix
```
Comment (rhyslbw) [inline nix/nixos/module-list.nix]:
```suggestion
  ./cardano-tx-submit-service.nix
```

**Comment by cleverca22** [comment]
bors try

File: .buildkite/docker-build-push.nix [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+#    the logged in Docker user's credentials.
+#
+# 2. The tag (default "VERSION") is changed to reflect the
+#    branch which is being built under this Buildkite pipeline.
+#
+#    - All commits are tagged with commit hash.
+#    - If the branch is master then master tag is updated
+#    - If there is a tag associated with the commit then that tag is pushed as well
+
+{ nodePackages ?  import ../. {}
```
Comment (disassembler) [inline .buildkite/docker-build-push.nix]: should be named something other than nodePackages restPackages or maybe just localPackages?

File: .buildkite/pipeline.yml [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+
+  - label: 'check-cabal-project'
+    command: 'nix-build ./nix -A iohkNix.checkCabalProject -o check-cabal-project.sh && ./check-cabal-project.sh'
+    agents:
+      system: x86_64-linux
+
+  - label: 'stack-cabal-sync'
+    command: 'nix-shell ./nix -A iohkNix.stack-cabal-sync-shell --run .buildkite/stack-cabal-sync.sh'
+    agents:
+      system: x86_64-linux
```
Comment (disassembler) [inline .buildkite/pipeline.yml]: needs to run the docker push script

File: default.nix [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    (selectProjectPackages cardanoRestHaskellPackages);
+
+  scripts = callPackage ./nix/scripts.nix { inherit customConfig; };
+
+  # NixOS tests
+  #nixosTests = import ./nix/nixos/tests {
+  #  inherit pkgs;
+  #};
+
+  dockerImage = let
```
Comment (disassembler) [inline default.nix]: should be plural since there are two?

**Comment by craigem** [comment]
bors r +

**Pull request merged**